rfc/rfc-42/rfc-42.md

```diff
@@ -0,0 +1,215 @@
+
+**For merging files**, we require that the buckets should be adjacent to each other in terms of their hash ranges so that the output bucket has only one continuous hash range.
+Though it is not required in a standard Consistent Hashing algorithm, fragmentations in hash ranges may cause extra complexity for the splitting process in our case.
+
+**For splitting files**, a split point (i.e., hash ranges for the output buckets) should be decided:
+
+- A simple policy would be split in the range middle, but it may produce uneven data files.
+In an extreme case, splitting may produce one file with all data and one file with no data.
+- Another policy is to find a split point that evenly dispatches records into children buckets.
+It requires knowledge about the hash value distribution of the original buckets.
+
+In our implementation, *we will first stick to the first simple one*, as buckets will finally converge to a balanced distribution after multiple rounds of resizing processes.
+Of course, a pluggable implementation will be kept for extensibility so that users can choose different available policies.
+
+All updates related to the hash metadata will be first recorded in the clustering plan, and then be reflected in partitions' hashing metadata when clustering finishes.
+The plan is generated and persisted in files during the scheduling process, which is protected by a table-level lock for a consistent table view.
+
+### Concurrent Writer & Reader
+
+Concurrent updates to file groups that are under clustering are not supported in Hudi, since records location change during the re-organizing.
+For example, a clustering process is scheduled to combine three parquet files A, B, C into two parquet files E, F with a more compact storage layout.
+During the clustering, a writer comes in and plan to update a record that originally belongs to the parquet file A.
+There are two choices the writer can make:
+
+- Only update the old file (i.e. parquet file A), but the update will lost once the clustering finishes because of the *replacecommit*'s shadowing.
+- Besides parquet file A, also update new files (i.e. parquet file E or F).
+However, the record's new location can only be determined until the clustering finishes, which means the writer will be blocked as it has to wait for the location information.
+
+For tables using Bucket Index, the above conflicting condition can be avoided because record locations are calculated by the hash algorithm.
+So even before the clustering finishes, the writer can calculate record locations as long as it knows the latest hash algorithm (in our Consistent Hashing case, it is stored in the clustering plan).
+
+![bucket resizing](bucket_resizing.png)
+
+The figure above shows a concurrent write process during the bucket resizing process (i.e., splitting).
+Dual-write is activated for buckets that are splitting.
```

Comment:
is this dual-write only happening when we generating the clustering plan?

Reply:
It happens when we are executing the clustering. As in single writer case, scheduling a clustering (i.e., generating the plan) is serialized to the writing.

Reply:
if we can block the writer when generating the clustering plan, do we still need to do the dual-write?

Reply:
Yes. The dual-write is necessary to allow asynchronous clustering. Regarding the generation of the plan, it is actually similar to compaction, where we also block writer while scheduling.

Reply:
let's say we are splitting bucket2 into bucket2.1 and bucket2.2, if we use dual-write that we will write bucket2.log and bucket2.1.log + bucket2.2.log. we can get the bucket2.log by merging bucket2.1.log + bucket2.2.log. Is that possible that we treat the two files as one? just like we treat the base file + log file as a base file before a pending compaction.

Reply:
oh ok, i got your point. Yes, writing log only once is the idealist case. Meanwhile, it requires modifying the read path to include the log merging logic from two other file groups. In current implementation, dual-write could keep the read path untouched, bringing down the implementation complexity. But in both way, we have to recognize ongoing resizing task in the write path (i.e., building a new hash view & re-tagging records), which i believe is the most expensive part.

By the way, I just came across an idea of `virtual log file`, i.e., 'empty' file referencing bucket2.1.log / bucket2.2.log. Similar to the bootstrapped file group, may be it could facilitate the read path by tracking small amount of necessary metadata.

Reply:
I didn't follow the recent work on clustering so I am not sure how's the async clustering feature going. If we can do async clustering just like async compaction, the read path of the files written after a clustering plan should have some rules to match with the files before the plan for MOR table. The bucket expansion is a perfect scenario for the async clustering. Is the async clustering(schedule inline and execute async) already supported?

Reply:
It is supported. But concurrent updates to file groups under clustering are not allowed.

Reply:
got it. it totally make sense to me. bucket expansion could be an exception cause we know the new file distribution so the concurrent updates should be possible.

Reply:
By the way, introducing dual-write, we actually want the write to be visible once it commits, even before the async clustering finishes. So only writing logs to 'child bucket' (i.e., bucket2.1 / bucket2.2) without modifying the read path, will delay the visibility (until the clustering finishes).

Reply:
yes, the ultimate solution is to support this through read path, or support this in the file group level. Treat the child buckets as one file group(like the virtual log file as you mentioned) before the clustering finish and separate them after the clustering committed would work. Cause we have the timeline information when constructing the file group. We have a similar logic for pending compaction I believe, like pretend a base file + a log file as a single base file. 
What I concerned was the dual-write might cause some extra operational workload and it's not streaming friendly. It's possible that the clustering plan would not be execute in a short amount of time.

Reply:
With resizing enabled, file group size will be controlled under a specified threshold (like max parquet file size). So in a normal case (reasonable initial bucket size and write speed), there would be an upper bound of the clustering execution time.

Reply:
Thanks for the discussion, Gray! I'll put the non-dual-write solution in the RFC.

Reply:
IIUC, the main point of the non-dual-write is to support reading two filegroup bucket2.1.log / bucket2.2.log + bucket2.parquet as bucket2, right? One thing of the rfc29 follow ups is that read data by bucket not file group to get better read performance by leverage hash distribution.

Reply:
The non-dual-write is to minimize the impact of resizing on the write performance. And it requires modification of the read path in order to understand the children buckets' logs. Reading based on hash distribution may be more complex in the context of consistent hashing, because partitions now have their own hashing metadata (we cannot align bucket id directly). I haven't dig into read path optimization yet, but I believe it is a great feature to put effects into.